File: 백준/Gold/17298. 오큰수/오큰수.py
import sys
read = sys.stdin.readline

N = int(read().rstrip())
ARR = list(map(int, read().rstrip().split(' ')))

length = len(ARR)
answer = [-1] * length
stack = [ARR[-1]]

for i in range(length-2, -1, -1):
  num = ARR[i]

  maxRight = stack[-1]
  # maxRight가 더 큰 경우
  if maxRight > num:
    answer[i] = maxRight
    stack.append(num)
    continue
  
  # num이 더 큰 경우
  while stack:
    maxRight = stack[-1]
    if maxRight > num: 
      answer[i] = maxRight
      break
    stack.pop()
  stack.append(num)

print(*answer)

File: 백준/Gold/17298. 오큰수/오큰수.js
const fs = require('fs');
const stdin = fs.readFileSync('/dev/stdin').toString().trim().split('\n');
const inputCnt = +stdin.shift();
const inputArr = stdin[0].trim().split(' ').map(Number);

function solution(cnt, arr) {
  const stack = [0];
  for (let i = 1; i < cnt; i++) {
    if (arr[i - 1] < arr[i]) {
      while (stack.length && arr[stack[stack.length - 1]] < arr[i]) {
        arr[stack.pop()] = arr[i];
      }
    }
    stack.push(i);
  }
    
  while (stack.length) {
    arr[stack.pop()] = -1;
  }
    
  console.log(arr.join(' '));
}

solution(inputCnt, inputArr);